libraries/physics/src/ThreadSafeDynamicsWorld.cpp

```diff
@@ -87,6 +87,33 @@ int ThreadSafeDynamicsWorld::stepSimulationWithSubstepCallback(btScalar timeStep
     return subSteps;
 }
 
+// call this instead of non-virtual btDiscreteDynamicsWorld::synchronizeSingleMotionState()
+void ThreadSafeDynamicsWorld::synchronizeMotionState(btRigidBody* body) {
+    btAssert(body);
+    if (body->getMotionState() && !body->isStaticObject()) {
+        //we need to call the update at least once, even for sleeping objects
+        //otherwise the 'graphics' transform never updates properly
+        ///@todo: add 'dirty' flag
+        //if (body->getActivationState() != ISLAND_SLEEPING)
+        {
+            if (body->isKinematicObject()) {
+                ObjectMotionState* objectMotionState = static_cast<ObjectMotionState*>(body->getMotionState());
+                if (!objectMotionState->hasInternalKinematicChanges()) {
+                    return;
+                } else {
+                    objectMotionState->clearInternalKinematicChanges();
+                }
+            }
+            btTransform interpolatedTransform;
+            btTransformUtil::integrateTransform(body->getInterpolationWorldTransform(),
+                body->getInterpolationLinearVelocity(),body->getInterpolationAngularVelocity(),
+                (m_latencyMotionStateInterpolation && m_fixedTimeStep) ? m_localTime - m_fixedTimeStep : m_localTime*body->getHitFraction(),
+                interpolatedTransform);
+            body->getMotionState()->setWorldTransform(interpolatedTransform);
+        }
+    }
+}
```

Comment:
This is a copy of btDiscreteDynamicsWorld::synchronizeSingleMotionState() (which is NOT virtual) with a change to keep an eye out for KINEMATIC objects that have been changed by Actions in the middle of the **stepSimulation()** (an unusual case for how Bullet was designed to work).